File: src/go/PhantasyIslandGoRemoteControl/http_layer.go
package PhantasyIslandGoRemoteControl

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// Ping 发送 ping 命令
func Ping() map[string]interface{} {
	return SendCmd("ping")
}

// PingVolatile 发送 ping 命令 (volatile)
func PingVolatile() map[string]interface{} {
	return SendCmdVolatile("ping")
}

// Start 发送 start 命令
func Start() map[string]interface{} {
	return SendCmd("start")
}

// StartVolatile 发送 start 命令 (volatile)
func StartVolatile() map[string]interface{} {
	return SendCmdVolatile("start")
}

// SendCmd 发送字符串命令
func SendCmd(s string) map[string]interface{} {
	client := http.Client{
		Timeout: 10 * time.Second,
	}
	url := fmt.Sprintf("http://%s/ECU_HTTP/sendStringCmd?c=%s", RemoteLocation, s)
	resp, err := client.Get(url)
	if err != nil {
		fmt.Printf("send_cmd %s Error Command Timeout or Connection Error: %v\n", s, err)
		return map[string]interface{}{"ok": false, "r": err.Error()}
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var j map[string]interface{}
	json.Unmarshal(body, &j)
	return j
}

// SendCmdVolatile 发送挥发性字符串命令
func SendCmdVolatile(s string) map[string]interface{} {
	client := http.Client{
		Timeout: 10 * time.Second,
	}
	url := fmt.Sprintf("http://%s/ECU_HTTP/sendStringCmd?cc=%s", RemoteLocation, s)
	resp, err := client.Get(url)
	if err != nil {
		fmt.Printf("send_cmd_volatile %s Error Command Timeout or Connection Error: %v\n", s, err)
		return map[string]interface{}{"ok": false, "r": err.Error()}
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var j map[string]interface{}
	json.Unmarshal(body, &j)
	return j
}

// GetAllAirplaneStatus 获取所有飞机状态
func GetAllAirplaneStatus() (map[string]interface{}, error) {
	client := http.Client{
		Timeout: 5 * time.Second,
	}
	url := fmt.Sprintf("http://%s/ECU_HTTP/requestPullAllAirplaneState", RemoteLocation)
	resp, err := client.Get(url)
	if err != nil {
		fmt.Fprintf(os.Stderr, "ConnectionError Cannot Connect to PhantasyIsland: %v\n", err)
		return nil, fmt.Errorf("ConnectionError Cannot Connect to PhantasyIsland, Max retries exceeded")
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var j map[string]interface{}
	json.Unmarshal(body, &j)
	return j, nil
}

// GetAirplaneCameraImage 获取指定无人机的摄像头图像
func GetAirplaneCameraImage(port string, camera string) *string {
	client := http.Client{
		Timeout: 5 * time.Second,
	}
	url := fmt.Sprintf("http://%s/ECU_HTTP/requestPullImage?flyPort=%s&imageType=%s", RemoteLocation, port, camera)
	resp, err := client.Get(url)
	if err != nil {
		fmt.Fprintf(os.Stderr, "ConnectionError Cannot Connect to PhantasyIsland: %v\n", err)
		return nil
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var j map[string]interface{}
	json.Unmarshal(body, &j)

	if ok, exists := j["ok"].(bool); exists && ok {
		if imgData, ok := j["imgDataString"].(string); ok {
			return &imgData
		}
	}
	return nil
}

// ProcessAirplane 处理飞机状态数据
func ProcessAirplane(j map[string]interface{}) map[string]map[string]interface{} {
	if ok, exists := j["ok"].(bool); exists && ok {
		airplanesRaw, exists := j["airplanes"].([]interface{})
		if !exists {
			return nil
		}

		airplaneStatus := make(map[string]map[string]interface{})
		for _, airRaw := range airplanesRaw {
			air := airRaw.(map[string]interface{})
			status := make(map[string]interface{})

			keyName := air["keyName"].(string)
			status["keyName"] = keyName
			status["typeName"] = air["typeName"]
			status["updateTimestamp"] = air["updateTimestamp"]
			status["status"] = air["status"]

			if cameraFront, ok := air["cameraFront"].(map[string]interface{}); ok {
				status["cameraFront"] = cameraFront["imgDataString"]
			}
			if cameraDown, ok := air["cameraDown"].(map[string]interface{}); ok {
				status["cameraDown"] = cameraDown["imgDataString"]
			}

			airplaneStatus[keyName] = status
		}
		return airplaneStatus
	}
	return nil
}

File: src/rs/src/http_layer.rs
use crate::config::REMOTE_LOCATION;
use std::collections::HashMap;
use serde_json::Value;

pub fn ping() -> Value {
    send_cmd("ping")
}

pub fn ping_volatile() -> Value {
    send_cmd_volatile("ping")
}

pub fn start() -> Value {
    send_cmd("start")
}

pub fn start_volatile() -> Value {
    send_cmd_volatile("start")
}

pub fn send_cmd(s: &str) -> Value {
    let url = format!("http://{}/ECU_HTTP/sendStringCmd?c={}", REMOTE_LOCATION, s);
    match reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .and_then(|c| c.get(&url).send()) {
        Ok(r) => r.json().unwrap_or(serde_json::json!({"ok": false, "r": "ParseError"})),
        Err(_) => serde_json::json!({"ok": false, "r": "Timeout or ConnectionError"}),
    }
}

pub fn send_cmd_volatile(s: &str) -> Value {
    let url = format!("http://{}/ECU_HTTP/sendStringCmd?cc={}", REMOTE_LOCATION, s);
    match reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .and_then(|c| c.get(&url).send()) {
        Ok(r) => r.json().unwrap_or(serde_json::json!({"ok": false, "r": "ParseError"})),
        Err(_) => serde_json::json!({"ok": false, "r": "Timeout or ConnectionError"}),
    }
}

pub fn get_all_airplane_status() -> Value {
    let url = format!("http://{}/ECU_HTTP/requestPullAllAirplaneState", REMOTE_LOCATION);
    match reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()
        .and_then(|c| c.get(&url).send()) {
        Ok(r) => r.json().unwrap_or(serde_json::json!({"ok": false, "r": "ParseError"})),
        Err(_) => serde_json::json!({"ok": false, "r": "ConnectionError"}),
    }
}

pub fn get_airplane_camera_image(port: &str, camera: &str) -> String {
    let url = format!("http://{}/ECU_HTTP/requestPullImage?flyPort={}&imageType={}", REMOTE_LOCATION, port, camera);
    match reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()
        .and_then(|c| c.get(&url).send()) {
        Ok(r) => {
            let j: Value = r.json().unwrap_or(serde_json::json!({"ok": false}));
            if j["ok"].as_bool().unwrap_or(false) {
                j["imgDataString"].as_str().unwrap_or("").to_string()
            } else {
                "".to_string()
            }
        },
        Err(_) => "".to_string(),
    }
}

pub fn process_airplane(j: Value) -> Option<HashMap<String, HashMap<String, Value>>> {
    if j["ok"].as_bool().unwrap_or(false) {
        let mut airplane_status = HashMap::new();
        if let Some(airplanes) = j["airplanes"].as_array() {
            for air in airplanes {
                let mut status = HashMap::new();
                let key_name = air["keyName"].as_str().unwrap_or("").to_string();
                status.insert("keyName".to_string(), Value::String(key_name.clone()));
                status.insert("typeName".to_string(), air["typeName"].clone());
                status.insert("updateTimestamp".to_string(), air["updateTimestamp"].clone());
                status.insert("status".to_string(), air["status"].clone());
                
                let camera_front = &air["cameraFront"];
                status.insert("cameraFront".to_string(), camera_front["imgDataString"].clone());
                
                let camera_down = &air["cameraDown"];
                status.insert("cameraDown".to_string(), camera_down["imgDataString"].clone());
                
                airplane_status.insert(key_name, status);
            }
        }
        Some(airplane_status)
    } else {
        None
    }
}
